hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -350,13 +365,86 @@ def span_end(self, discard: bool, /) -> None:  # noqa: B027, FBT001
         """
 
 
+def _permitted_constants(
+    constants: Any,
+    *,
+    choice_type: ChoiceTypeT,
+    constraints: ChoiceConstraintsT,
+) -> tuple["ConstantT", ...]:
+    return tuple(
+        value
+        for value in constants
+        if choice_permitted(
+            int_to_float(value) if choice_type == "float" else value,
+            constraints,
+        )
+    )
+
+
 class HypothesisProvider(PrimitiveProvider):
     lifetime = "test_case"
 
     def __init__(self, conjecturedata: Optional["ConjectureData"], /):
         super().__init__(conjecturedata)
+        self.local_constants = _get_local_constants()
         self._random = None if self._cd is None else self._cd._random
 
+    def _maybe_draw_constant(
+        self,
+        choice_type: ChoiceTypeT,
+        constraints: ChoiceConstraintsT,
+        *,
+        p: float = 0.05,
```

Comment:
I think I'd put this outside the method: if we're not actually using the constants then I'd rather not pay to fill the cache.

Reply:
It probably won't matter much either way, a particular set of `ChoiceConstraintsT` that gets used for long enough is going to hit the 5% chance of drawing a constant and fill the cache then. I don't mind doing this though, at worst it's equivalent and at best it saves some space.